arm/cstart64.S

```diff
@@ -71,19 +71,42 @@ start:
 	add	x1, x1, :lo12:stacktop
 	bl	setup				// x0 is the addr of the dtb
 
+	b	run_test
+#endif
+
+.text
+
+/*
+ * Used by cpu0 to setup the system and start the actual test.
+ * x0 is the address of the dtb
+ * x1 is the start of the free memory
+ * x2 is the top of the stack
+*/
+.global primary_entry
+primary_entry:
+	/* set up stack */
+	mov	x4, #0
+	msr	spsel, x4
+	isb
+	sub	x2, x2, 16			// stackptr must be strictly
+						// less than stacktop
+	mov	sp, x2
+
+	bl	setup
+
+	b	run_test
+
+run_test:
 	/* run the test */
 	adrp	x0, __argc
-	ldr	x0, [x0, :lo12:__argc]
+	ldrsw	x0, [x0, :lo12:__argc]
```

Comment:
I was worried  about the upper bits of x0, but since the result should be used as an int, it shouldn't be an issue. main will look at w0 not x0 so this should work.